File: Computational Mathematics/countmpi.c
/*      program countmpi.c */
/* Illustrates count for arrays. */      
#include <stdio.h>
/* Includes the mpi C library. */
#include "mpi.h"
      main(int argc, char* argv[]) {
      float a[4];
      int b[2][3];
      int my_rank,p,n,i,j,dest,tag;
      n    = 4;
      dest = 0;
      tag  = 50;
/* Initializes mpi, gets the rank of the processor, my_rank, */
/* and number of processors, p. */
      MPI_Init(&argc, &argv);
      MPI_Comm_rank(MPI_COMM_WORLD,&my_rank);
      MPI_Comm_size(MPI_COMM_WORLD,&p);
/* Define the arrays. */     
      if (my_rank==.0) {
        a[0] = 1.;
        a[1] = 2.78;
        a[2] = 3.14;
        a[3] = 186000.;
        for (j = 0;j<3;j++) {
          for (i = 0;i<2;i++) {
            b[i][j] = i+j;
          }
        }
      }
/* Each processor attempts to print the array. */    
      for (i =0;i<4;i++) {
         printf("my_rank = %d a(%d) = %f\n",my_rank,i,a[i]);
      } 
      MPI_Barrier(MPI_COMM_WORLD);
/* The arrays are broadcast via count equal to four. */     
      MPI_Bcast(&(a[0]),4,MPI_FLOAT,0,
                        MPI_COMM_WORLD);
      MPI_Bcast(&(b[0][0]),4,MPI_INT,0,
                        MPI_COMM_WORLD);                  
/* Each processor prints the arrays. */ 
      for (i =0;i<4;i++) {
         printf("my_rank = %d a(%d) = %f\n",my_rank,i,a[i]); 
      }                     
      for (j = 0;j<3;j++) {
         for (i = 0;i<2;i++) {
            printf("my_rank = %d b(%d,%d) = %d\n",my_rank,i,j,b[i][j]);
         }
      } 
/* mpi is terminated. */            
      MPI_Finalize();
}  /*   end program countmpi */

File: Computational Mathematics/countmpi.f
      program countmpi
! Illustrates count for arrays.      
      implicit none
      include 'mpif.h'
      real, dimension(1:4):: a
      integer, dimension(1:2,1:3):: b
      integer::  my_rank,p,n,source,dest,tag,ierr,loc_n
      integer:: i,j,status(mpi_status_size)
      data n,dest,tag/4,0,50/
      call mpi_init(ierr)
      call mpi_comm_rank(mpi_comm_world,my_rank,ierr)
      call mpi_comm_size(mpi_comm_world,p,ierr)
! Define the arrays.      
      if (my_rank.eq.0) then
        a(1) = 1.
        a(2) = exp(1.)
        a(3) = 4*atan(1.)
        a(4) = 186000.
        do j = 1,3
          do i = 1,2
            b(i,j) = i+j
          end do
        end do
      end if
! Each processor attempts to print the array.     
      print*,'my_rank =',my_rank, 'a = ',a
      call mpi_barrier(mpi_comm_world,ierr)
! The arrays are broadcast via count equal to four.      
      call mpi_bcast(a,4,mpi_real,0,&
                        mpi_comm_world,ierr)
      call mpi_bcast(b,4,mpi_int,0,&
                        mpi_comm_world,ierr)                  
! Each processor prints the arrays.                        
      print*,'my_rank =',my_rank, 'a = ',a
      print*,'my_rank =',my_rank, 'b = ',b              
      call mpi_finalize(ierr)
      end program countmpi

